lib/services.js

```diff
@@ -737,4 +769,131 @@ export default {
             metadata.deleteObjectMD(mpuBucketName, key, {}, log, callback);
         }, err => cb(err));
     },
+
+    versioningPreprocessing(bucketName, bucketMD, objectKey, objMD,
+            reqVersionId, log, callback) {
+        const options = {};
+        // bucket is not versioning enabled
+        if (!bucketMD.getVersioningConfiguration()) {
+            options.deleteData = true;
+            return callback(null, options);
+        }
+        // bucket is versioning enabled
+        const mstVersionId = objMD ? objMD.versionId : undefined;
+        const mstIsNull = objMD ? objMD.isNull : false;
+        const vstat = bucketMD.getVersioningConfiguration().Status;
+        if (!reqVersionId) {
+            // non-version-specific versioning operation
+            if (mstVersionId === undefined || mstIsNull) {
+                // object does not exist or is not versioned (before versioning)
+                if (vstat === 'Suspended') {
+                    // versioning is suspended, overwrite the existing version
+                    options.versionId = '';
+                    options.isNull = true;
+                    options.deleteData = true;
+                    return callback(null, options);
+                }
+                // versioning is enabled, create a new version
+                options.versioning = true;
+                if (objMD) {
+                    // store master version in a new key
+                    const versionId = mstIsNull ? mstVersionId : VID_INF;
+                    objMD.versionId = versionId; // eslint-disable-line
+                    objMD.isNull = true; // eslint-disable-line
+                    options.nullVersionId = versionId;
+                    return metadata.putObjectMD(bucketName, objectKey, objMD,
+                            { versionId }, log, err => callback(err, options));
+                }
+                return callback(null, options);
+            }
+            // master is versioned and is not a null version
+            const nullVersionId = objMD.nullVersionId;
+            if (vstat === 'Suspended') {
+                // versioning is suspended, overwrite the existing version
+                options.versionId = '';
+                options.isNull = true;
+                options.deleteNullVersionData = true;
+                if (nullVersionId === undefined) {
```

Comment:
@ttvi FYI: I think we need to bump this if condition up one line (we should not attempt to delete null version data if there is no null version)